python/src/iceberg/avro/decoder.py

```diff
@@ -0,0 +1,167 @@
+        if len(read_bytes) != n:
+            raise ValueError(f"Read {len(read_bytes)} bytes, expected {n} bytes")
+        return read_bytes
+
+    def read_boolean(self) -> bool:
+        """
+        a boolean is written as a single byte
+        whose value is either 0 (false) or 1 (true).
+        """
+        return ord(self.read(1)) == 1
+
+    def read_int(self) -> int:
+        """int values are written using variable-length, zigzag coding."""
+        return self.read_long()
+
+    def read_long(self) -> int:
+        """long values are written using variable-length, zigzag coding."""
+        b = ord(self.read(1))
+        n = b & 0x7F
+        shift = 7
+        while (b & 0x80) != 0:
+            b = ord(self.read(1))
+            n |= (b & 0x7F) << shift
+            shift += 7
+        datum = (n >> 1) ^ -(n & 1)
+        return datum
+
+    def read_float(self) -> float:
+        """
+        A float is written as 4 bytes.
+        The float is converted into a 32-bit integer using a method equivalent to
+        Java's floatToIntBits and then encoded in little-endian format.
+        """
+        return float(STRUCT_FLOAT.unpack(self.read(4))[0])
+
+    def read_double(self) -> float:
+        """
+        A double is written as 8 bytes.
+        The double is converted into a 64-bit integer using a method equivalent to
+        Java's doubleToLongBits and then encoded in little-endian format.
+        """
+        return float(STRUCT_DOUBLE.unpack(self.read(8))[0])
+
+    def read_decimal_from_bytes(self, precision: int, scale: int) -> decimal.Decimal:
+        """
+        Decimal bytes are decoded as signed short, int or long depending on the
+        size of bytes.
+        """
+        size = self.read_long()
+        return self.read_decimal_from_fixed(precision, scale, size)
+
+    def read_decimal_from_fixed(self, precision: int, scale: int, size: int) -> decimal.Decimal:
+        """
+        Decimal is encoded as fixed. Fixed instances are encoded using the
+        number of bytes declared in the schema.
+        """
+        data = self.read(size)
+        unscaled_datum = int.from_bytes(data, byteorder="big", signed=True)
+        return unscaled_to_decimal(unscaled_datum, scale)
+
+    def read_bytes(self) -> bytes:
+        """
+        Bytes are encoded as a long followed by that many bytes of data.
+        """
+        return self.read(self.read_long())
+
+    def read_utf8(self) -> str:
+        """
+        A string is encoded as a long followed by
+        that many bytes of UTF-8 encoded character data.
+        """
+        return self.read_bytes().decode("utf-8")
+
+    def read_date_from_int(self) -> date:
+        """
+        int is decoded as python date object.
+        int stores the number of days from
+        the unix epoch, 1 January 1970 (ISO calendar).
+        """
+        days_since_epoch = self.read_int()
+        return date(1970, 1, 1) + timedelta(days_since_epoch)
```

Comment:
Minor: I'd prefer to use `days_to_date` just like this uses `micros_to_time` and `micros_to_timestamp`.

Reply:
I like it 👍🏻  Updated